### PR TITLE
[FLINK-3871] Add Kafka TableSource with Avro serialization

### PR DESCRIPTION
**Pull request by mushketyk**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by fhueske** [comment]
Hi @mushketyk, thanks for this pull request! 
I noticed, it addresses two separate issues: TableSources (FLINK-3871) and TableSinks (FLINK-3873) for Avro-encoded Kafka topics. Would you mind, splitting up the PR into two? This will make the review a bit faster. 

Thanks, Fabian

**Comment by mushketyk** [comment]
Hi @fhueske 

Thank you for pointing this out. I didn't know that there are two separate issues for that.
I can split this PR into two and put serializer, deserializer and table sources in one PR and sinks into another one.
Alternatively I can split this PR into 3 PRs: 
* serializer and deserializer
* sources
* sinks

Which split would you prefer?

Best regards,
Ivan.

**Comment by fhueske** [comment]
I'd go with the first approach, i.e., two PRs. You should reference the PR on which the second one refers.

Thanks, Fabian

**Comment by mushketyk** [comment]
@fhueske Thank you for clarification. I'll update the code tonight.

**Comment by mushketyk** [comment]
@fhueske I've split the PR according to your suggestion. Now this PR only contains serializer, deserializer and table source.
Since sink part depends on this PR I cannot publish it now, so I'll publish it when this PR is merged.

**Comment by fhueske** [comment]
Thanks a lot for the update @mushketyk. I'm quite busy at the moment but will try to have a look at your PR soon. Thanks, Fabian

**Comment by mushketyk** [comment]
@fhueske No worries. Take your time.

Best regards,
Ivan.

**Comment by fhueske** [comment]
Hi @mushketyk, do you plan to continue with this PR?

Thanks, Fabian

**Comment by mushketyk** [comment]
HI @fhueske, sorry I was a bit busy lately.  I'll update the PR tonight.

Best regards, 
Ivan

**Comment by fhueske** [comment]
No worries and no need to hurry! :-)
Thanks, Fabian

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		String[] fieldNames,
+		TypeInformation<?>[] fieldTypes) {
+
+		super(topic, properties, createDeserializationSchema(fieldTypes), fieldNames, fieldTypes);
+	}
+
+	private static AvroRowDeserializationSchema createDeserializationSchema(
+		TypeInformation<?>[] fieldTypes) {
+
+		return new AvroRowDeserializationSchema(new String[]{"f1", "f2", "f3"}, fieldTypes);
```
Comment (mushketyk) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java]: Sorry, accidentally typed it.
Will fix.

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	private static AvroRowDeserializationSchema createDeserializationSchema(
+		Class<?>[] fieldTypes) {
+
+		return new AvroRowDeserializationSchema(new String[]{"f1", "f2", "f3"}, TypeUtil.toTypeInfo(fieldTypes));
```
Comment (mushketyk) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java]: Ditto.

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowSerializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	/** Field names in a Row */
+	private final String[] fieldNames;
+	/** Avro serialization schema */
+	private final Schema schema;
+	/** Writer to serialize Avro GeneralRecord into a byte array */
+	private final DatumWriter<GenericRecord> datumWriter;
+	/** Output stream to serialize records into byte array */
+	private final ByteArrayOutputStream arrayOutputStream =  new ByteArrayOutputStream();
+	/** Low level class for serialization of Avro values */
+	private final Encoder encoder = EncoderFactory.get().directBinaryEncoder(arrayOutputStream, null);
```
Comment (mushketyk) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowSerializationSchema.java]: Ok, thank you for suggestion.

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowDeserializationSchema.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	private final TypeInformation[] fieldTypes;
+	/** Avro deserialization schema */
+	private final Schema schema;
+	/** Reader that deserializes byte array into a record */
+	private final DatumReader<GenericRecord> datumReader;
+	/** Record to deserialize byte array to */
+	private final GenericRecord record;
+
+	/**
+	 * Creates a Avro deserializtion schema for the given type classes.
```
Comment (mushketyk) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowDeserializationSchema.java]: Good catch. Thank you.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/AvroDeserializationSchemaTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		byte[] bytes = serializationSchema.serialize(row);
+		deserializationSchema.deserialize(bytes);
+		deserializationSchema.deserialize(bytes);
+		Row resultRow = deserializationSchema.deserialize(bytes);
+
+		assertEqualsRows(row, resultRow);
+	}
+
+	private Row createRow() {
+		Row row = new Row(3);
```
Comment (mushketyk) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/AvroDeserializationSchemaTest.java]: Makes sense. I'll add these tests.

**Comment by mushketyk** [comment]
@fhueske  Just a small question about your comment. You wrote:
"users should be able to provide a Schema and a String ->String to convert aGenericRecordinto a Row"

What is String -> String for?

**Comment by fhueske** [comment]
The `String -> String` mapping would be a name mapping of fields from the `GenericRecord` to `Row` attributes. This would allow users to freely choose the attribute names of a table which is generated from an Avro schema.

**Comment by mushketyk** [comment]
Are there any other cases when a user needs a custom conversion from a `GenericRecord` to a `Row` instance?
`String -> String` will definitely work, but I also can think about a generic convertor `GenericRecord => Row`, but this may be an overkill. What do you think about this?

**Comment by fhueske** [comment]
Would be good if the TableSource could be used by providing parameters and without extending it.

**Comment by mushketyk** [comment]
Hi @fhueske 
Ok, I'll give a try.

I've pushed an updated PR where Schema is now external. I'll work on the transformation part tomorrow.

I had few issues while trying to accommodate your comments.
1. When I try to to use DateTime as a value of one of the fields of a GenericRecord I receive the following exception:

```
0    [main] INFO  org.apache.flink.api.java.typeutils.TypeExtractor  - class org.joda.time.DateTime is not a valid POJO type

java.lang.RuntimeException: java.lang.InstantiationException

	at org.apache.avro.specific.SpecificData.newInstance(SpecificData.java:316)
	at org.apache.avro.specific.SpecificData.newRecord(SpecificData.java:332)
	at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:173)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:151)
	at org.apache.avro.generic.GenericDatumReader.readField(GenericDatumReader.java:193)
	at org.apache.avro.reflect.ReflectDatumReader.readField(ReflectDatumReader.java:230)
	at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:183)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:151)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:155)
	at org.apache.avro.generic.GenericDatumReader.readField(GenericDatumReader.java:193)
	at org.apache.avro.reflect.ReflectDatumReader.readField(ReflectDatumReader.java:230)
	at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:183)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:151)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:142)
	at org.apache.flink.streaming.util.serialization.AvroRowDeserializationSchema.readRecord(AvroRowDeserializationSchema.java:77)
	at org.apache.flink.streaming.util.serialization.AvroRowDeserializationSchema.deserialize(AvroRowDeserializationSchema.java:70)
	at org.apache.flink.streaming.connectors.kafka.AvroDeserializationSchemaTest.deserializeRowWithComplexTypes(AvroDeserializationSchemaTest.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.lang.InstantiationException
	at sun.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.apache.avro.specific.SpecificData.newInstance(SpecificData.java:314)
	... 43 more
```

I am not how to serialize DateTime using Avro. Do you have any suggestions?

2. You suggested that "creating a new ByteArrayInputStream and Decoder for each record is quite expensive". I don't see a way how to do this. The problem is that to create Decoder I need to pass an instance of an InputStream, but when an ByteArrayInputStream is created its content cannot be changed. Therefore a new instance of ByteArrayInputStream should be created for each message.

I think I can create a different implementation of InputStream that allows to set a buffer (let's call it `MutableByteArrayInputStream`). Then we can create one instance of `MutableByteArrayInputStream` and just call `inputStream.setBuffer(message)` for every message that should be deserialized.

**Comment by mushketyk** [comment]
Hi @fhueske 

I've updated PR according to your review.

Best regards,
Ivan.

**Comment by fhueske** [comment]
Hi @mushketyk, while reviewing your PR I noticed a significant shortcoming of the `TableSource` interface. `TableSource` does not support nested schemas. IMO, this needs to be fixed before we can have a proper integration of Avro integration with the Table API / SQL. 

I created [FLINK-5280](https://issues.apache.org/jira/browse/FLINK-5280) and also updated the description of this issue ([FLINK-3871](https://issues.apache.org/jira/browse/FLINK-3871). The `KafkaJsonTableSource` needs to be reworked to support nested data as well ([FLINK-5281](https://issues.apache.org/jira/browse/FLINK-5281)).

I think we have to pause this PR until FLINK-5280 is resolved.
What do you think?

Best, Fabian

**Comment by mushketyk** [comment]
Hi @fhueske ,

Sorry, I am still relatively new to Flink so I don't have the full context. Could you elaborate why this is an issue and how does this shortcoming affect this PR?

Do you have an idea how difficult will it be to implement this change? If it's not very involved I could help with it to move this PR forward.

Best regards,
Ivan.

**Comment by fhueske** [comment]
Hi @mushketyk, sorry for the concise description before. The problem is the following:

A `TableSource` provides schema information for the Table it produces. However, the methods `TableSource.getFieldNames()` and `TableSouce.getFieldTypes()` return flat arrays which are interpreted as flat schema (the second field in the array represents the second table attribute) without any nesting.

Avro and many other storage formats (JSON, Parquet, ...) support nested data structures. With the current limitation of the `TableSource` interface, we would need to convert the nested data into a flat schema. However, the Table API and SQL support processing of nested data and it would be a much better integration to pass Avro objects in their original structure into the Table API / SQL query (see my updated proposal for [FLINK-3871](https://issues.apache.org/jira/browse/FLINK-3871). 

In order to be able to create a Table of nested Avro data, we need to improve the `TableSource` interface first. Once this is done, we can continue with this PR.

I'm very sorry, that I did not think about this earlier and the effort you already put into this issue.
Please let me know if you have any questions. 
Best, Fabian

**Comment by twalthr** [comment]
@mushketyk if you don't mind, I would continue working on this issue.

**Pull request closed without merging**